### PR TITLE
fence regions 

### PR DESCRIPTION
**Pull request by Aurora-73**
一、
PlaceDB.py", line 190, in scale
    self.flat_region_boxes -= box_shift_factor
修改为
    self.flat_region_boxes -= np.repeat(box_shift_factor, len(self.regions))
修改原因：box_shift_factor是一个有四个元素的一维数组，但是flat_region_boxes包含不止一个region，长度应该是4 * N

二、
PlaceDB.py", line 195, in scale
    self.regions[i] -= box_shift_factor
修改为
    self.regions[i] -= np.mean(box_shift_factor)

三、
ops/fence_region/fence_region.py", line 127
新增
    if regions.dim() == 1:
        regions = regions.unsqueeze(0)

四、
PlaceDB.py", line 718, in calc_num_filler_for_fence_region
    placeable_area = np.sum((region[:, 2] - region[:, 0]) * (region[:, 3] - region[:, 1]))
修改为
    placeable_area = np.sum((region[2] - region[0]) * (region[3] - region[1]))
修改原因：region已经是regions的一个子数组了，是一维数组

五、
PlaceDB.py", line 701, in calc_num_filler_for_fence_region
        lower_bound = np.percentile(movable_node_size_x, 5)
        upper_bound = np.percentile(movable_node_size_x, 95)
        filler_size_x = np.mean(
            movable_node_size_x[(movable_node_size_x >= lower_bound) & (movable_node_size_x <= upper_bound)]
        )
修改为
        if len(movable_node_size_x) > 0:
            lower_bound = np.percentile(movable_node_size_x, 5)
            upper_bound = np.percentile(movable_node_size_x, 95)
            filler_size_x = np.mean(
                movable_node_size_x[(movable_node_size_x >= lower_bound) & (movable_node_size_x <= upper_bound)]
            )
        else:
            lower_bound = 0
            upper_bound = 0
            filler_size_x = 0
修改原因：有可能存在不含任意元素的空region（如果所有的node都指定了所属region会出现这种情况，最后会多一个不含任何node的region）

六、
PlaceDB.py", line 731, in calc_num_filler_for_fence_region
    fence_regions = np.concatenate(self.regions, 0).astype(np.float32)
修改为
    fence_regions = np.array(self.regions)

七、
PlaceDB.py", line 752, in calc_num_filler_for_fence_region
        num_filler = int(round(total_filler_node_area / (filler_size_x * filler_size_y)))
修改为
        if filler_size_x > 0 and filler_size_y > 0:
            num_filler = int(round(total_filler_node_area / (filler_size_x * filler_size_y)))
        else:
            num_filler = 0
修改原因：同<四>，(filler_size_x * filler_size_y)可能为0
    
八、
BasicPlace.py", line 295, in __init__
新增
    region = np.atleast_2d(region)

九、
BasicPlace.py", line 790, in build_fence_region_legalization
新增
    virtual_macros = np.atleast_2d(virtual_macros)

十、
BasicPlace.py", line 793, in build_fence_region_legalization
        virtual_macros_size_x = (virtual_macros[:,2]-virtual_macros[:,0]).clamp(min=30)

        virtual_macros_size_y = (virtual_macros[:,3]-virtual_macros[:,1]).clamp(min=30)       
修改为
        virtual_macros_size_x = np.clip(virtual_macros[:, 2] - virtual_macros[:, 0], a_min=30, a_max=None)
        virtual_macros_size_y = np.clip(virtual_macros[:, 3] - virtual_macros[:, 1], a_min=30, a_max=None)

十一、
BasicPlace.py", line 797, in build_fence_region_legalization
        virtual_macros_pos = virtual_macros[:,0:2].t().contiguous()
修改为
        virtual_macros_tensor = torch.from_numpy(virtual_macros)
        virtual_macros_pos = virtual_macros_tensor[:,0:2].t().contiguous()

十二、
BasicPlace.py", line 803, in build_fence_region_legalization
新增
        virtual_macros_size_x = torch.tensor(virtual_macros_size_x, dtype=torch.float32, device=device)
        virtual_macros_size_y = torch.tensor(virtual_macros_size_y, dtype=torch.float32, device=device)

十三、
electric_potential.py", line 397, in compute_fence_region_map
新增
        if fence_region.dim() == 1:
            fence_region = fence_region.view(-1, 4)

十四、
NonLinearPlace.py", line 103, in __call__
    density_weight = all_metrics[-1][-1][-1].density_weight.item() / params.two_stage_density_scaler
修改为
    density_weight = np.array(all_metrics[-1][-1][-1].density_weight.tolist()).mean() / params.two_stage_density_scaler

十五、
NonLinearPlace.py", line 234, in Lgamma_stop_criterion
    if len(placedb.regions) > 0 and model.update_mask.sum() == 0:
修改为
    if (len(placedb.regions) > 0 and model.update_mask == None) or (len(placedb.regions) > 0 and model.update_mask.sum() == 0):
修改原因：同<四>，model.update_mask可能为空

### CODE REVIEW

**Review by magic3007 — COMMENTED** [review]
1. NumPy has a broadcasting mechanism that can handle this situation. Here is an example of NumPy's broadcasting mechanism:
```
>>> a
array([[0, 1, 2, 3],
       [4, 5, 6, 7]])
>>> b
array([0, 1, 2, 3])
>>> a - b
array([[0, 0, 0, 0],
       [4, 4, 4, 4]])
```
2. I don't understand why we need to use `mean`.

3-15. Did you encounter these problems when there was only one region?
Note that **a region constraint** can consist of multiple rectangles.
For example, the `region` variable in Fix (4) actually describes multiple rectangles within a region constraint.
